### PR TITLE
A new matcher for iproute

### PR DESCRIPTION
**Pull request by razum2um**
In order to dig into the library and `pallet` I developed another server matcher which could check traffic routes.

Backstory, if you curious why: there is kind of "Great Russian Firewall" managed by Roskomnadzor in Russia. They still blocked part of AWS(notably gems.contribsys.com), LinkedIn(+Slideshare) and many others. So I'm leveraging 2 interfaces on my router: provider uplink and vpn. There is a *huge* script which effectively does: 

```
ip route add default via 10.8.0.32 table ovpnc1
FOR EACH <blocked-ip> DO rule add to <blocked-ip> table ovpnc1 END
```

This script works quite slow, besides if I need to update it - they current way is only clear tables and re-fill. Besides, switching to full vpn => slow bandwith :\

TLDR: I did things manually and thought of a tool which can monitor current list and probably will add provisioning soon to resolve ips and inject missing `ip rule`

Feel free to close without merge if you think it's useless for others :)

### CODE REVIEW

**Comment by razum2um** [comment]
Specs will fail because it needs https://github.com/DomainDrivenArchitecture/dda-config-commons/pull/6

**Comment by jerger** [comment]
like your idea :-)

File: project.clj [hunk 1]
```diff
@@ -3,8 +3,10 @@
   :url "https://domaindrivenarchitecture.org"
   :license {:name "Apache License, Version 2.0"
             :url "https://www.apache.org/licenses/LICENSE-2.0.html"}
-  :dependencies [[dda/dda-pallet-commons "1.4.0"]
-                 [dda/dda-pallet "2.2.1"]]
+  :dependencies [[dda/dda-pallet-commons "1.4.1-SNAPSHOT"]
+                 [dda/dda-config-commons "1.4.2-SNAPSHOT"]
+                 [dda/dda-pallet "2.2.1"]
+                 [iproute "0.1.4"]]
```
Comment (razum2um) [inline project.clj]: feeling strange about this 😐and made some thoughts about extendability...

**Comment by jerger** [comment]
yes we will do this in front of releasing the dda-serverspec-crate

File: project.clj (hunk 1 above)
Comment (jerger) [reply project.clj]: we will release commons as soon as PR is working.

File: main/src/dda/pallet/dda_serverspec_crate/domain.clj [hunk 2]
```diff
@@ -44,6 +45,8 @@
                                         :expiration-days s/Num}]  ;min days certificate must be valid
    (s/optional-key :http) [{:url s/Str                            ;full url e.g. http://google.com
                             :expiration-days s/Num}]              ;minimum days the certificate must be valid
+   (s/optional-key :iproute) [{:ip (s/pred schema/ipv4?)
+                              :via (s/pred schema/ipv4?)}]
```
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_serverspec_crate/domain.clj]: I've thought about this one more time:

1) preamble: ipv6 cannot be another key, since we cannot mix bewteen 6 and 4. actually, further schema will relax to `:ip v4-or-v6, :via v4-or-v6` + both should be the same version. (if :via is v6 hostname will be resolved to 6). So, according to Rich Hickey it's non-breaking enhancement, so I'd leave the keys unchanged. 

> completely different
it's not, instead of `ip route get` we'll have `ip -6 route get` but I have troubles to setup and proper test ipv6 in my ISP environment :(

TLDR:  I suggest to leave as-is

2) device & source can be added but I'd rather rework *all* infra.test.* nses because they're too tedious while checking facts (extracting values, building error message in ad-hoc way). we have an explicit step for `domain -> infra` maybe we need a reverse step which ensures that extracted facts converted to the *same shape* as domain schema (in iproute case it's effectively same shape already) and we'll be able to use some generic hash-diff algorithm to check expectations. 

Ideally I'd except something like this concise output with clear distinctions: https://camo.githubusercontent.com/b2ab86a13264be23b1eac3a10aa8c8199c0ccb92/687474703a2f2f662e636c2e6c792f6974656d732f33413155313232443144327731753368324f30682f53637265656e25323053686f74253230323031332d31322d33312532306174253230392e30302e3331253230414d2e706e67 

TLDR: device & source will require ad-hoc comparison code in infra.test.iproute which I'd ideally generalize before

[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_serverspec_crate/domain.clj]: btw, good visual hash-matcher-expectations in clj are https://github.com/weavejester/eftest and https://github.com/pjstadig/humane-test-output rspec just came first to my mind :)

File: main/src/dda/pallet/dda_serverspec_crate/infra/fact/iproute.clj [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+(s/defn parse-iproute-response :- IprouteFactResult
+  "returns a IprouteFactResult from the result of a gateway probe"
+  [single-script-result :- s/Str]
+  (let [result-lines (string/split single-script-result #"\n" 2)
+        result-key (first result-lines)
+        result-text (nth result-lines 1)]
+    ;; iproute always tries to find multiple rules
+    {(keyword result-key) (-> (route/parse result-text)
+                              first
```
[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_serverspec_crate/infra/fact/iproute.clj]: Thanks for notice, anyway: I remembered that I stripped output line with "cached" but then added the multiline support to `iproute` lib itself + included sample output: https://github.com/DomainDrivenArchitecture/dda-serverspec-crate/pull/15/commits/bfac0e029ef404d9ebbbd9cb262bcb512d9927b4

File: project.clj [hunk 4]
```diff
@@ -4,7 +4,8 @@
   :license {:name "Apache License, Version 2.0"
             :url "https://www.apache.org/licenses/LICENSE-2.0.html"}
   :dependencies [[dda/dda-pallet-commons "1.4.0"]
-                 [dda/dda-pallet "2.2.1"]]
+                 [dda/dda-pallet "2.2.1"]
+                 [iproute "0.1.4"]]
```
[the remark this answers is not in the document]
Comment (razum2um) [reply project.clj]: First of all, even if we count all transitive *runtime* deps (ignoring profiles), then there are *a lot* of EPL thing there (even without org.clojure/*) already: See https://gist.github.com/razum2um/1e0fd52a2e7f8925de8169fa61cc761c 

So `iproute` & `instaparse` (which btw has no runtime deps expect clojure; criterium & rhizome are `:dev`) are "too late to concern about". EPL is highly promoted by "lein new" imo ;)

Anyway, as I told you non-gpl stuff is pretty interchangeable, also see: https://opensource.stackexchange.com/questions/5843/can-i-distribute-my-software-using-a-license-other-than-epl-if-i-use-an-epl-depe

> Instaparse has a (hard?) clojure 1.8 dependency

actually a good point. but afair lein builds correct classpath with proper clojure version notwithstanding on which versions depend various dependencies (and covers uberjar as well leaving single proper clojure version inside).

File: main/src/dda/pallet/dda_serverspec_crate/domain.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_serverspec_crate/domain.clj]: Reg: better visual matching - better readable output is a good idea.

[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_serverspec_crate/domain.clj]: Reg. ip4 vs ip6: `{:ip (s/pred schema/ipv4?)` this will obvious only work for v4 - so name should reflect this fact.
Due to ensure the "non-breaking change in future" contract we should wise choose names in order to make our live easy ;-)

At the dependency between `:ip` and `:via` your right - but there are ways to express this in schema as in spec (for the future ...)

But lets widen our view, why not chosing:
```
(s/optional-key :iproute4) [{:ip (s/pred schema/ipv4?)
                                           :via (s/pred schema/ipv4?)}]
```

[the remark this answers is not in the document]
Comment (razum2um) [reply main/src/dda/pallet/dda_serverspec_crate/domain.clj]: > this will obvious only work for v4

Sorry, I think the core misunderstanding happened due I believe that names designed for a *very near* feature (even maybe *partially* supported) are *ok*, and I both disliked `:ipv4` and `:iproute4` in config since they are freezing bad `if`s inside code (final future result is imo *worse*)

But code is better than words, so I resigned and implemented both `ipv6`,`dev`/`src` matching and better colored comparison with `clojure.data/diff` http://take.ms/dVhBF

Let's leave `:ip` name, ok :)?

[the remark this answers is not in the document]
Comment (jerger) [reply main/src/dda/pallet/dda_serverspec_crate/domain.clj]: reviwed & fine :-)

**Comment by jerger** [comment]
can u update the reference in doc folder?
